### ISSUE TITLE
`FuseReshape`: the collapsed extent is not equivalent to the rank-3 bounds in four cases

### ISSUE THREAD

**Issue by etiotto** [intel/intel-xpu-backend-for-triton#8001]
`FuseReshape` collapses two descriptor dimensions into one index `i*d + j` and replaces the per-dimension bounds checks with a single check on the merged extent (`FuseReshape.cpp:175-195`). `isCandidate` (`FuseReshape.cpp:281-321`) does not reject the following. None are reachable from the `_grouped_mm` template, so they are separate from #7924.

All results: `main` @ `9fa71f9d5`, Arc B580.

## 1. Unaligned merged-dim offset reads rows the rank-3 descriptor pads

`isCandidate` never inspects the load offset. When `off_merged + blockExtent > s1` and `off_collapsed < s0-1`, rows `j >= s1` satisfy `q = off0*d + j < (s0-1)*d + s1` and are read instead of padded.

```python
import torch, triton, triton.language as tl
triton.set_allocator(lambda s, a, st: torch.empty(s, device="xpu", dtype=torch.int8))

@triton.jit
def k(a_ptr, i_ptr, o_ptr):
    d = tl.make_tensor_descriptor(a_ptr, shape=[4, 128, 64], strides=[8192, 64, 1],
                                  block_shape=[1, 32, 64])
    a = d.load([0, 97, 0]).reshape(32, 64)   # rows 97..128; row 128 is outside dim1 == 128
    i = tl.load(i_ptr + tl.arange(0, 64)[:, None] * 64 + tl.arange(0, 64)[None, :])
    tl.store(o_ptr + tl.arange(0, 32)[:, None] * 64 + tl.arange(0, 64)[None, :], tl.dot(a, i))

buf = torch.zeros(4 * 128 * 64, device="xpu", dtype=torch.bfloat16)
buf.view(4, 128, 64)[1, 0, :] = 3.0          # slice 1 row 0 -- not part of the loaded slice
out = torch.zeros(32, 64, device="xpu", dtype=torch.float32)
k[(1,)](buf, torch.eye(64, device="xpu", dtype=torch.bfloat16), out)
print(out[31, :4].tolist())
```

Expected `[0.0, 0.0, 0.0, 0.0]`, got `[3.0, 3.0, 3.0, 3.0]`. The value tracks the fill. Silent wrong data, no error. `off_collapsed == s0-1` is unaffected.

Same defect on the middle-dim branch (`collapsedDim == 1`): shape `[4,8,64]`, strides `[512,64,1]`, block `[32,1,64]`, offsets `[0,0,1]` -- column `k=64` is read against `width=512`.

## 2. Negative outer stride fuses and returns padding

```python
# as above, but:
    d = tl.make_tensor_descriptor(a_ptr, shape=[4, 128, 64], strides=[-8192, 64, 1],
                                  block_shape=[1, 32, 64])
    a = d.load([1, 0, 0]).reshape(32, 64)

for p in range(4):
    buf.view(4, 128, 64)[p, :, :] = float(p + 1)
k[(1,)](buf[3 * 128 * 64:], torch.eye(64, device="xpu", dtype=torch.bfloat16), out)
```

Base at the last slice with `strides[0] < 0` is a reverse-ordered batch, so logical slice 1 is physical slice 2. Expected `3.0`, got `0.0` -- the collapsed row coordinate goes negative and the whole block reads as padding. Setting `shape=[4,100,64]` declines fusion and returns `3.0` correctly.

`isDivisible` accepts the negative stride via `getZExtValue()` (`third_party/intel/lib/TritonIntelGPUTransforms/Utility.cpp:67`).

## 3. `shapes[collapsedDim] == 0` wraps the extent on i32

`(shapes[collapsedDim] - 1)` is computed on i32 (`FuseReshape.cpp:192-194`).

```mlir
tt.func public @f(%arg0: tensor<256x32xbf16>, %arg1: !tt.ptr<bf16>) -> tensor<256x256xf32> {
  %s0 = arith.constant 0 : i32
  %s1 = arith.constant 128 : i32
  %s2 = arith.constant 256 : i32
  %a = arith.constant 65536 : i64
  %b = arith.constant 256 : i64
  %c = arith.constant 1 : i64
  %off = arith.constant 0 : i32
  %cst = arith.constant dense<0.000000e+00> : tensor<256x256xf32>
  %0 = tt.make_tensor_descriptor %arg1, [%s0, %s1, %s2], [%a, %b, %c] : <bf16>, <1x32x256xbf16>
  %1 = tt.descriptor_load %0[%off, %off, %off] : !tt.tensordesc<1x32x256xbf16> -> tensor<1x32x256xbf16>
  %2 = tt.reshape %1 : tensor<1x32x256xbf16> -> tensor<32x256xbf16>
  %3 = tt.dot %arg0, %2, %cst, inputPrecision = tf32 : tensor<256x32xbf16> * tensor<32x256xbf16> -> tensor<256x256xf32>
  tt.return %3 : tensor<256x256xf32>
}
```

`triton-opt -triton-intel-fuse-reshape -canonicalize` gives declared height `-128 : i32`, i.e. **4294967168** as an unsigned surface height:

```
%0 = tt.make_tensor_descriptor %arg1, [%c-128_i32, %c256_i32], [%c256_i64, %c1_i64] : <bf16>, <32x256xbf16>
```

Only when `d > s1`; `d == s1` gives 0 and `d < s1` under-declares. On the middle-dim branch with `s1 = 0` the same wrap lands in the surface width (`b = 2*s2` -> 4294967232). The 24-bit range check at `LowerTo2DBlockLoad.cpp:283-288` covers only the inner shape and the pitch, not `shapes[descRank-2]` (`:298`).

## 4. Middle-dim collapse can declare a surface wider than its pitch

The merged extent becomes the surface width while `strides[0]` stays the pitch.

```mlir
tt.func public @f(%arg0: tensor<64x256xbf16>, %arg1: !tt.ptr<bf16>) -> tensor<32x256xf32> {
  %s0 = arith.constant 4 : i32
  %s1 = arith.constant 8 : i32
  %s2 = arith.constant 64 : i32
  %a = arith.constant 64 : i64
  %b = arith.constant 64 : i64
  %c = arith.constant 1 : i64
  %off = arith.constant 0 : i32
  %cst = arith.constant dense<0.000000e+00> : tensor<32x256xf32>
  %0 = tt.make_tensor_descriptor %arg1, [%s0, %s1, %s2], [%a, %b, %c] : <bf16>, <32x1x64xbf16>
  %1 = tt.descriptor_load %0[%off, %off, %off] : !tt.tensordesc<32x1x64xbf16> -> tensor<32x1x64xbf16>
  %2 = tt.reshape %1 : tensor<32x1x64xbf16> -> tensor<32x64xbf16>
  %3 = tt.dot %2, %arg0, %cst, inputPrecision = tf32 : tensor<32x64xbf16> * tensor<64x256xbf16> -> tensor<32x256xf32>
  tt.return %3 : tensor<32x256xf32>
}
```

Same command gives width **512** against pitch **64**:

```
%0 = tt.make_tensor_descriptor %arg1, [%c4_i32, %c512_i32], [%c64_i64, %c1_i64] : <bf16>, <32x64xbf16>
```

**Comment by etiotto** [comment]
Verification notes for PR #8069, moved out of the PR description to keep it focused.

### Alignment-compensation measurement (B580)

The middle-branch width guard is `extentBytes <= pitchBytes`. `width == pitch` was
not obviously the true legality boundary: when the base needs alignment
compensation, lowering aligns it down and folds the misalignment into the width
while leaving the pitch alone, so an effective width can exceed its pitch.

Measured rather than assumed. A device-side descriptor with `width == pitch == 128`
bytes, a non-foldable column offset of 48, and a base deliberately misaligned by
16 bytes (`base % 64 == 16`), loaded through a `tl.dot` so it reaches the DPAS 2D
path:

- `ttg.threads-per-warp = 16`, `ttig.2d_block_io_base_alignment = 64`, so
  compensation is live
- the `tt.descriptor_load` carries `ttig.block_io = "row_major"` and a
  `dot_op`/`mma` layout; the LLIR has 4 `spirv_Subgroup2DBlockLoad` calls
- the out-of-bounds columns read **0**, not the following row's data (which is
  physically present at those addresses, so the check is sensitive)

It pads. Nothing to tighten.


### Regression runs (B580)

- `python/test/unit/intel/test_host_tensor_descriptor.py` — 8 passed (4 pre-existing + 4 added here)
- `python/test/unit/intel/test_block_tdesc.py`, `test_block_io.py` — 443 passed
- `python/test/unit/language/test_tensor_descriptor.py` — 1560 passed, 1322 xfailed,
  3 failed: `test_tensor_descriptor_load_nd[16-2-1-{float32,int32,uint32}]`.
  Those 3 fail identically with this branch's pass reverted, so they are
  pre-existing and unrelated. They are not in any skiplist. Measured on a host
  that needs a local SPIR-V extension substitution to run `tl.dot` kernels at
  all, so CI is the authority on whether they reproduce there.

`block_io` canary: the fused rank-2 `tt.descriptor_load` still carries
`ttig.block_io = "row_major"` in the TTGIR, and the merged extent is still
`muli(addi(muli(clamp, ratio), shape), extui(cmpi sgt))` — a top-level `muli`,
not a `select`, which is what keeps `ttgi::isDivisible` able to prove what
`MaterializeBlockPointer` needs.

### Follow-ups filed (deliberately not fixed here)

- #8070 — per-dimension *lower* bounds are still not preserved (pre-existing)
- #8071 — `LowerTo2DBlockLoad` never range-checks the surface height
- #8072 — non-foldable negative strides/shapes are still trusted, which is why
  the negative-stride fix is limited to foldable strides
- #8073 — `ttgi::isDivisible` takes an `unsigned` divisor and divides by it
  unguarded
